cmd/sensors/src/lib.rs

```diff
@@ -302,6 +327,179 @@ fn print(
             errs = vec![None; rval.len()];
         }
 
+        Ok(SensorData { values: rval, errors: errs })
+    }
+}
+
+/// Reader for sensor data that pulls it out of RAM directly
+struct RamSensorReader {
+    /// Position of `LAST_READING` in global RAM
+    last_reading: HubrisVariable,
+    last_reading_buf: Vec<u8>,
+
+    /// Position of `DATA_VALUE` in global RAM
+    data_value: HubrisVariable,
+    data_value_buf: Vec<u8>,
+
+    /// Position of `NERRORS` in global RAM
+    nerrors: HubrisVariable,
+    nerrors_buf: Vec<u8>,
+
+    /// Indexes of sensors that we care about
+    ///
+    /// We'll read the entire array (for simplicity) then filter just the ones
+    /// that matter
+    sensors: Vec<usize>,
+}
+
+impl RamSensorReader {
+    fn new(
+        hubris: &HubrisArchive,
+        sensors: &[(usize, HubrisSensor)],
+    ) -> Result<Self> {
+        let find_var = |name| {
+            hubris
+                .qualified_variables()
+                .find(|&(n, _)| n == format!("task_sensor::main::{name}"))
+                .ok_or_else(|| anyhow!("could not find {name}"))
+                .map(|(_, v)| *v)
+        };
+        let last_reading = find_var("LAST_READING")?;
+        let data_value = find_var("DATA_VALUE")?;
+        let nerrors = find_var("NERRORS")?;
+
+        Ok(Self {
+            last_reading,
+            last_reading_buf: vec![0u8; last_reading.size],
+            data_value,
+            data_value_buf: vec![0u8; data_value.size],
+            nerrors,
+            nerrors_buf: vec![0u8; nerrors.size],
+            sensors: sensors.iter().map(|(i, _)| *i).collect(),
+        })
+    }
+
+    fn unpack_array<T, F: Fn(&Value) -> Result<T>>(
+        &self,
+        hubris: &HubrisArchive,
+        buf: &[u8],
+        var: HubrisVariable,
+        f: F,
+    ) -> Result<Vec<T>> {
+        let vs = humility::reflect::load_value(
+            hubris,
+            buf,
+            hubris.lookup_type(var.goff)?,
+            0,
+        )?;
+        let Value::Struct(vs) = vs else {
```

Comment:
So it seems like a _lot_ of this code could be simplified by describing the structure you want as a `struct` and applying `derive(Load)`. You could write the struct inline here. It'd also (IMO) make it clearer to a reader what things you're depending on.

I did this recently in the host boot reporting code if you want an example: https://github.com/oxidecomputer/humility/pull/492/files#diff-5290acdb0fe5054172bdefac251ae764fa71a2e5a7e938f1b43af61cbd7e1f5fR88

Reply:
Done in ff6df5c20fcfb55be866017c89011348c1d189e9 ; it's not quite a LOC increase, because I had to add `doppel::MaybeUninit<T>`, but now that's around for the future.